core/src/backend/audio.rs

```diff
@@ -560,16 +557,19 @@ pub struct SoundTransform {
     pub right_to_right: f32,
 }
 
-impl SoundTransform {
+impl From<display_object::SoundTransform> for SoundTransform {
     /// Converts from a `display_object::SoundTransform` to a `backend::audio::SoundTransform`.
-    fn from_display_object_transform(other: &DisplayObjectSoundTransform) -> Self {
-        const SCALE: f32 = (display_object::SoundTransform::MAX_VOLUME
-            * display_object::SoundTransform::MAX_VOLUME) as f32;
+    fn from(other: display_object::SoundTransform) -> Self {
+        const SCALE: f32 = display_object::SoundTransform::MAX_VOLUME.pow(2) as f32;
+
+        // It seems like Flash stores sound transform values in 30-bit unsigned integers:
+        // * In general, negative values are equivalent to their absolute value.
+        // * Specifically, 0x40000000, -0x40000000 and -0x80000000 are equivalent to zero.
         Self {
-            left_to_left: other.left_to_left as f32 * other.volume as f32 / SCALE,
-            left_to_right: other.left_to_right as f32 * other.volume as f32 / SCALE,
-            right_to_left: other.right_to_left as f32 * other.volume as f32 / SCALE,
-            right_to_right: other.right_to_right as f32 * other.volume as f32 / SCALE,
+            left_to_left: ((other.left_to_left << 2) >> 2) as f32 * other.volume as f32 / SCALE,
+            left_to_right: ((other.left_to_right << 2) >> 2) as f32 * other.volume as f32 / SCALE,
+            right_to_left: ((other.right_to_left << 2) >> 2) as f32 * other.volume as f32 / SCALE,
+            right_to_right: ((other.right_to_right << 2) >> 2) as f32 * other.volume as f32 / SCALE,
```

Comment:
Weird behavior!

I think the multiply should be integer using `wrapping_mul`, something like:
`left_to_left: ((other.left_to_left.wrapping_mul(other.volume) << 2) >> 2) as f32 / SCALE,`
Here's an example that this fixes, but not 100% sure. 512 * 4096 = 2^32 -> wraps to 0
```
var s = new Sound(mc);
s.attachSound("marching");

var t = {
	ll: 4096,
	lr: 0,
	rl: 0,
	rr: 0
};
s.setVolume(1048576);
s.setTransform(t);
var t = s.getTransform();
trace("ll: " + t.ll);
trace("lr: " + t.lr);
trace("rl: " + t.rl);
trace("rr: " + t.rr);
trace(s.getVolume());
s.start();

```

[sound_vol_wrap.zip](https://github.com/ruffle-rs/ruffle/files/7117848/sound_vol_wrap.zip)